test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs

```diff
@@ -19,6 +19,866 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
+                    },
+                    {
+                        "<div style=\"@DateTime.Now\" class=\"@DateTime.Now\" custom=\"@DateTime.Now\" >words and " +
+                        "spaces</div>",
+                        new MarkupBlock(
+                            new MarkupTagHelperBlock(
+                                "div",
+                                attributes: new Dictionary<string, SyntaxTreeNode>
+                                {
+                                    ["style"] = dateTimeNow,
+                                    ["class"] = dateTimeNow,
+                                    ["custom"] = dateTimeNow
+                                },
+                                children: factory.Markup("words and spaces")))
+                    },
+                    {
+                        "<div style=\"\" class=\"btn\" custom=\"hi\" >words<strong>and</strong>spaces</div>",
+                        new MarkupBlock(
+                            new MarkupTagHelperBlock(
+                                "div",
+                                attributes: new Dictionary<string, SyntaxTreeNode>
+                                {
+                                    ["style"] = new MarkupBlock(),
+                                    ["class"] = factory.Markup("btn"),
+                                    ["custom"] = factory.Markup("hi")
+                                },
+                                children: new SyntaxTreeNode[]
+                                {
+                                    factory.Markup("words"),
+                                    blockFactory.MarkupTagBlock("<strong>"),
+                                    factory.Markup("and"),
+                                    blockFactory.MarkupTagBlock("</strong>"),
+                                    factory.Markup("spaces")
+                                }))
+                    },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(RequiredAttributeData))]
+        public void Rewrite_RequiredAttributeDescriptorsCreateTagHelperBlocksCorrectly(
+            string documentContent,
+            MarkupBlock expectedOutput)
+        {
+            // Arrange
+            var descriptors = new TagHelperDescriptor[]
+                {
+                    new TagHelperDescriptor(
+                        tagName: "p",
+                        typeName: "pTagHelper",
+                        assemblyName: "SomeAssembly",
+                        attributes: new TagHelperAttributeDescriptor[0],
+                        requiredAttributes: new[] { "class" }),
+                    new TagHelperDescriptor(
+                        tagName: "div",
+                        typeName: "divTagHelper",
+                        assemblyName: "SomeAssembly",
+                        attributes: new TagHelperAttributeDescriptor[0],
+                        requiredAttributes: new[] { "class", "style" }),
```

Comment:
nit: the theory data would be easier to grok if you were using explicit names like `requiredForP`, `requiredForDiv1`, and `requiredForDiv2`.  (figured it out but...)